antenna/crashmover.py

```diff
@@ -137,102 +87,47 @@ def check_health(self, state):
-                    self.crashmover_publish(crash_report)
-                    self.crashmover_finish(crash_report)
-
+                self.crashmover_save(crash_report)
+                break
             except Exception:
-                MYMETRICS.incr("%s_crash_exception.count" % crash_report.state)
-                crash_report.errors += 1
+                MYMETRICS.incr("save_crash_exception.count")
                 LOGGER.exception(
-                    "Exception when processing queue (%s), state: %s; error %d/%d",
-                    crash_report.crash_id,
-                    crash_report.state,
-                    crash_report.errors,
-                    MAX_ATTEMPTS,
+                    f"Exception when saving crash ({crash_id}); error {errors}/{MAX_ATTEMPTS}"
                 )
+        else:
+            # After MAX_ATTEMPTS, we give up on this crash
+            LOGGER.error(f"{crash_id}: too many errors trying to save; dropped")
+            MYMETRICS.incr("save_crash_dropped.count")
+            return False
 
-                # After MAX_ATTEMPTS, we give up on this crash and move on
-                if crash_report.errors < MAX_ATTEMPTS:
-                    self.crashmover_queue.append(crash_report)
-                else:
-                    LOGGER.error(
-                        "%s: too many errors trying to %s; dropped",
-                        crash_report.crash_id,
-                        crash_report.state,
-                    )
-                    MYMETRICS.incr("%s_crash_dropped.count" % crash_report.state)
-
-    def crashmover_finish(self, crash_report):
-        """Finish bookkeeping on crash report."""
-        # Capture the total time it took for this crash to be handled from
-        # being received from breakpad client to saving to s3.
-        #
-        # NOTE(willkg): We use submitted_timestamp which is formatted with isoformat().
-        # time.time() returns seconds as a float, but .timing() wants milliseconds, so
-        # we multiply by 1000.
-
-        delta = time.time() - isoformat_to_time(
-            crash_report.raw_crash["submitted_timestamp"]
-        )
-        delta = delta * 1000
+        for errors in range(1, MAX_ATTEMPTS + 1):
+            try:
+                self.crashmover_publish(crash_report)
+                MYMETRICS.incr("save_crash.count")
+                return True
+            except Exception:
+                MYMETRICS.incr("publish_crash_exception.count")
+                LOGGER.exception(
+                    f"Exception when publishing crash ({crash_id}); error {errors}/{MAX_ATTEMPTS}"
+                )
 
-        MYMETRICS.timing("crash_handling.time", value=delta)
-        MYMETRICS.incr("save_crash.count")
+        # After MAX_ATTEMPTS, we give up on this crash
+        LOGGER.error(f"{crash_id}: too many errors trying to publish; dropped")
+        MYMETRICS.incr("publish_crash_dropped.count")
+        return False
```

Comment:
If we've gotten here, then the crash report data is in S3 and all we need to do is publish the crash id to the processing queue. If this fails, Socorro will notice. It's got a daily task that looks for raw crash files for which there's no corresponding processed crash file and add an entry to the missing processed crashes table. You can see that here:

https://crash-stats.mozilla.org/siteadmin/crashstats/missingprocessedcrash/

That page lets you reprocess crash ids where there isn't a processed crash report.

I thought Socorro also self-healed in that it would automatically reprocessed crash reports, but it doesn't. I forget what I was thinking in that regard. I must have been fixing it by hand for a time. There's a `MissingProcessedCrashes` API and a corresponding `fetch_missing` command, so that seems likely.

* https://crash-stats.mozilla.org/api/#MissingProcessedCrash
* https://github.com/mozilla-services/socorro/blob/main/socorro/scripts/fetch_missing.py

We should fix that to self-heal--it should put crash reports in the queue to get processed. If processing fails, it'll send Sentry events, so we'll always have some signal of crash reports failing to process. I'll write up a bug for that. It should be a straight-forward fix to `verifyprocessed.py`.

Anyhow, that's a long train of thought to get to what Antenna should do when it succeeds in saving the crash data to S3, but fails at publishing the crash id for processing. In this case, I think `handle_crashreport` should return `True` and Antenna should return an HTTP 200 with the crash id. Socorro should catch the problem and self-heal.

Also, the way this loop is written, it'll retry 20 times in rapid succession. I think it should retry 5 times waiting 2 seconds in between attempts and then give up and return `True`.